File: proj-c/src/answers.c
#include <gmodule.h>
#include <stdlib.h>

#include "answers.h"

/**
@file answers.c
Estrutura de dados relativa às respostas, respetivos
getters e setters e funções auxiliares necessárias.
*/

/**
\brief Estrutura dos posts respostas.
*/
typedef struct answers {
  long user_id;
  long answer_id;
  long parent_id; // a que pergunta pertence
  int score;
  int comment_count; //número de comentários recebidos pela resposta (query 10)
} answers;

/**
\brief Função que devolve o tamanho da estrutura answers.
@returns int - Tamanho da estrutura answers.
*/
int sizeAnswers() {
    return sizeof(struct answers);
}

/**
\brief Função que devolve o id do user que deu a resposta.
@param a Apontador para answers.
@returns long - Título e utilizador.
*/
long getAUserId(Answers a) {
    return a->user_id;
}

/**
\brief Função que estabelece o id do user que deu determinada resposta.
@param a Apontador para answers.
@param id Identificador do user.
*/
void setAUserId(Answers a, long id) {
    a->user_id = id;
}

/**
\brief Função que devolve o id da resposta.
@param a Apontador para answers.
@returns long - Identificador da resposta.
*/
long getAnswerId(Answers a) {
    return a->answer_id;
}

/**
\brief Função que estabelece o id da resposta.
@param a Apontador para answers.
@param id Identificador da resposta.
*/
void setAnswerId(Answers a, long id) {
    a->answer_id = id;
}

/**
\brief Função que devolve o id da pergunta a que a resposta se refere.
@param a Apontador para answers.
@returns long - Identificador da pergunta a que a resposta está associada.
*/
long getParentId(Answers a) {
    return a->parent_id;
}

/**
\brief Função que estabelece o id da pergunta a que a resposta se refere.
@param a Apontador para answers.
@param id Identificador da pergunta a que a resposta está associada.
*/
void setParentId(Answers a, long id) {
    a->parent_id = id;
}

/**
\brief Função que devolve o score de uma resposta.
@param a Apontador para answers.
@returns int - Score.
*/
int getScore(Answers a) {
    return a->score;
}

/**
\brief Função que estabelece o score de determinada resposta.
@param a Apontador para answers.
@param s Score.
*/
void setScore(Answers a, int s) {
    a->score = s;
}

/**
\brief Função que devolve o número de comentários de uma resposta.
@param a Apontador para answers.
@returns int - Número de Comentários.
*/
int getCommentCount(Answers a) {
    return a->comment_count;
}

/**
\brief Função que estabelece o número de comentários de uma resposta.\brief
@param a Apontador para answers.
@param c Número de comentários.
*/
void setCommentCount(Answers a, int c) {
    a->comment_count = c;
}

/**
\brief Função que compara os votos de duas respostas
(query 6).
@param a1 Apontador para apontador para answers.
@param a2 Apontador para apontador para answers.
@returns gint - Positivo se a1 < a2 e negativo caso contrário.
*/
static gint sortVotes(Answers * a1, Answers * a2) {
  int votes_a1 = 0, votes_a2 = 0;
  votes_a1 = getScore(*a1);
  votes_a2 = getScore(*a2);

    if(votes_a1 < votes_a2)
      return 1;
    else
      return -1;
}


/**
\brief Função que ordena os votos das respostas por ordem decrescente
do número de votos (query 6).
@param total_answers Apontador para um GPtrArray.
*/
void sortAnswersVotes(GPtrArray * total_answers) {
  g_ptr_array_sort(total_answers, (GCompareFunc)sortVotes);
}

/**
\brief Função que liberta o apontador para a estrutura answers.
@param a Apontador para answers.
*/
void cleanAnswer(Answers a) {
    free(a);
}

File: proj-java/LI3/Answers.java

/**
 * Escreva a descrição da classe Answers aqui.
 *

 * @author A81047
 * @author A34900
 * @author A82339
 * @version 20180519
 */

public class Answers extends Posts {
  private long parent_id; // a que pergunta pertence
  private int score;
  private int comment_count; //número de comentários recebidos pela resposta (query 10)

  public Answers() {
      super();
      parent_id = 0;
      score = 0;
      comment_count = 0;
    }

  public Answers(long user_id, long post_id, long parent_id, int score, int comment_count) {
    super(user_id, post_id);
    this.parent_id = parent_id;
    this.score = score;
    this.comment_count = comment_count;
    }

  public Answers(Answers a) {
      super(a);
      this.parent_id = a.getParentId();
      this.score = a.getScore();
      this.comment_count = a.getCommentCount();
    }

  /**
     * Função que devolve o id da pergunta a que a resposta pertence.
     *
     * @returns long - Identificador da pergunta a que a resposta pertence.
     */
    public long getParentId() {
      return parent_id;
    }

  /**
     * Função que estabelece o id da pergunta a que a resposta pertence.
     *
     * @param id Identificador da pergunta a que a resposta pertence.
     */
    public void setParentId(long id) {
      parent_id = id;
    }

  /**
     * Função que devolve o score da resposta
     *
     * @returns int - score
     */
    public int getScore() {
      return score;
    }

  /**
     * Função que estabelece o score da resposta
     *
     * @param score Score da resposta
     */
    public void setScore(int score) {
      this.score = score;
    }

  /**
     * Função que devolve o número de comentários da resposta.
     *
     * @returns int - número de comentários
     */
    public int getCommentCount() {
      return comment_count;
    }

  /**
     * Função que estabelece o número de comentários da resposta.
     *
     * @param comment_count o número de comentários da resposta.
     */
    public void setCommentCount(int comment_count) {
      this.comment_count = comment_count;
    }

  public Answers clone() {
     return new Answers(this);
  }

  //TODO equals e toString
}
